### PR TITLE
Add HTML best practices

### PR DESCRIPTION
**Pull request by ericwbailey**
This PR expands on guidance for HTML best practices.

### CODE REVIEW

File: best-practices/README.md [hunk 1]
```diff
@@ -216,7 +216,26 @@ JavaScript
 HTML
 ----
 
-* Use `<button>` tags over `<a>` tags for actions.
+* Use lowercase text for elements and attributes
+* Use double quotes to wrap element attributes
+* Use closing tags for all elements. Implicitly closed elements can be 
+  unpredictable and have unintended consequences
+* Prefer a HTML5 doctype and utf-8 charset
```
Comment (tysongach) [inline best-practices/README.md]:
```suggestion
* Prefer a HTML5 doctype and UTF-8 charset
```

…or…

```suggestion
* Prefer a HTML5 doctype and `utf-8` charset
```

Comment (tysongach) [inline best-practices/README.md]:
```suggestion
* Prefer an HTML5 doctype and utf-8 charset
```

Comment (tysongach) [inline best-practices/README.md]:
```suggestion
* Prefer a HTML5 doctype and utf-8 character set
```

…or…

```suggestion
* Prefer a HTML5 doctype and utf-8 `charset`
```

File: best-practices/README.md [hunk 2]
```diff
@@ -216,7 +216,26 @@ JavaScript
...
+* Ensure elements intended to be used in either the `<head>` and `<body>` 
+  elements are scoped properly
```
Comment (tysongach) [inline best-practices/README.md]: I’m not sure I clearly understand what this means…

File: best-practices/README.md [hunk 3]
```diff
@@ -216,7 +216,26 @@ JavaScript
...
+* Ensure `id`s are unique
+* Append relevant attribute values to an existing attribute value, if present
```
Comment (tysongach) [inline best-practices/README.md]: I’m not sure I clearly understand what this means…

File: best-practices/README.md [hunk 4]
```diff
@@ -216,7 +216,26 @@ JavaScript
...
+* Avoid using emoji and other exotic characters as attribute values
```
Comment (tysongach) [inline best-practices/README.md]: `yes="👏"`

**Comment by ericwbailey** [comment]
Took a shot at clarifying the two bullet points you flagged. Do they read better to you now?

File: best-practices/README.md [hunk 5]
```diff
@@ -216,7 +216,32 @@ JavaScript
 HTML
 ----
 
-* Use `<button>` tags over `<a>` tags for actions.
+* Use lowercase text for elements and attributes
+* Use double quotes to wrap element attributes
+* Use closing tags for all elements. Implicitly closed elements can be 
+  unpredictable and have unintended consequences
```
[the remark this answers is not in the document]
Comment (ericwbailey) [reply best-practices/README.md]: This is more for [elements that can auto-close in theory](https://meiert.com/en/blog/optional-html/), but isn't so great in practice.

Do you think we should also have XHTML-style? I do it personally, but it's more habit than anything else at this point.

File: best-practices/README.md [hunk 6]
```diff
@@ -216,7 +216,32 @@ JavaScript
...
+* Prefer a HTML5 doctype and UTF-8 charset
```
[the remark this answers is not in the document]
Comment (ericwbailey) [reply best-practices/README.md]: I've always seen it included as defensive boilerplate. To be honest, I'm not sure of any edge cases if the header is doing its job, but that also doesn't mean they don't exist?

I do think messaging about using UTF-8 should be included somewhere.

File: best-practices/README.md [hunk 7]
```diff
@@ -216,7 +216,32 @@ JavaScript
...
+  * Elements such as `<title>` and `<meta>` should be placed within the page's 
+    `<head>` element
+  * Elements such as `<p>`, `<nav>`, `<div>`, etc. should be placed within the 
+    page's `<body>` element
+* Ensure `id`s are unique
+* Prefer appending attribute values instead of declaring redunant attribute names
+  * For example, if adding a class of `c-card--featured`, add it to the existing
+    class declaration (`class="c-card c-card--featured"`, not 
+    `class="c-card" class="c-card--featured"`)
+* Avoid using emoji and other exotic characters as attribute values
```
[the remark this answers is not in the document]
Comment (ericwbailey) [reply best-practices/README.md]: Adrian Roselli has a good wrapup of [a lot of the concerns](http://adrianroselli.com/2017/10/avoid-emoji-as-class-names.html). I've also seen them trip up corporate content filters before.

File: best-practices/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (mike-burns) [reply best-practices/README.md]: I don't use XHTML-style, and I also don't ever type e.g. `</br>` or `</link>`. I skimmed that article but might not have understood -- is this setting myself up for failure?

(I did use XHTML style when I wrote XHTML but stopped when I went back to HTML 4.01.)

File: best-practices/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mike-burns) [reply best-practices/README.md]: Maybe in the "Web" section?

File: best-practices/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mike-burns) [reply best-practices/README.md]: Oh I see. I was worried there was a problem with e.g. `<input type="hidden" name="_utf8" value="☃️">`, but that article doesn't imply that we should worry about that. This is more about `class`, `id`, and ARIA values?

[the remark this answers is not in the document]
Comment (ericwbailey) [reply best-practices/README.md]: Good point about the [snowman](http://unicodesnowmanforyou.com/) trick, I'll update this point to speak more to what attributes the point is speaking to.

File: best-practices/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ericwbailey) [reply best-practices/README.md]: Works for me!

File: best-practices/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (ericwbailey) [reply best-practices/README.md]: You're good on not using XHTML-style. The potential problem this point is addressing is elements that used to require a closing element that a browser can technically close, but it is still better to be explicit about the scope of what the element is wrapping by writing a closing tag.

So, something like:

```html
<ul>
  <li><a href="">Cats</a>
  <li><a href="">Dogs</a>
</ul>
```

will technically parse in most modern browsers, but runs the risk of getting malformed if more markup is added to the pattern. 

I'd also argue that it's extra cognitive overhead to remember which elements can optionally be auto-closed and which can't.

[the remark this answers is not in the document]
Comment (mike-burns) [reply best-practices/README.md]: Ah right, that. I think the distinction the spec makes is [void](https://html.spec.whatwg.org/multipage/syntax.html#void-elements) vs non-void.

Maybe

> Explicitly close all non-void elements.

Conveys the point both specifically and usefully?

[the remark this answers is not in the document]
Comment (ericwbailey) [reply best-practices/README.md]: I think that's great, thank you! I'll update it alongside hour other suggestions.

**Comment by ericwbailey** [comment]
@mike-burns Okay to merge this?

**Comment by mike-burns** [comment]
LGTM. Thank you!

**Comment by tysongach** [comment]
👏 👏 👏

**Pull request merged**